### PR TITLE
GG-37304 .NET: Add UnwrapNullablePrimitiveTypes binary configuration mode

### PR DESCRIPTION
**Pull request by ptupitsyn**
**Problem**

GridGain.NET (in any mode - thin client, thick client, server) writes object fields of nullable primitive types (such as `int?`) using `BinaryWriter.WriteObject`, which is inconsistent with Java behavior, where `BinaryFieldAccessor` uses non-public methods in `BinaryWriterExImpl` to write nullable primitives, producing different metadata.

**Solution**

Add `BinaryConfiguration.UnwrapNullablePrimitiveTypes` flag which changes the behavior to match Java
* .NET specific
* Works for thin client and embedded modes the same way
* Off by default (compatibility), but recommended to enable when possible
* Log a warning when not enabled

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/platform/PlatformNullablePrimitivesTask.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (isapego) [inline modules/core/src/test/java/org/apache/ignite/platform/PlatformNullablePrimitivesTask.java:2]: I believe, we should put actual date in new files?
```suggestion
 * Copyright 2023 GridGain Systems, Inc. and Contributors.
```

File: modules/platforms/dotnet/Apache.Ignite.Core/Ignition.cs [hunk 2]
```diff
@@ -367,6 +367,27 @@ private static void LogDiagnosticMessages(IgniteConfiguration cfg, ILogger log)
...
+        {
+            if (log == null)
+                return;
+
+            if (cfg == null || !cfg.UnwrapNullablePrimitiveTypes)
+            {
+                log.Warn(
+                    $"{nameof(BinaryConfiguration)}.{nameof(BinaryConfiguration.UnwrapNullablePrimitiveTypes)} " +
+                    "is not enabled. It is recommended to enable this setting, " +
+                    "unless old behavior is required for compatibility reasons.");
```
Comment (isapego) [inline modules/platforms/dotnet/Apache.Ignite.Core/Ignition.cs:390]: I'm not sure guys from CS would approve on this. Better check with @DirectXceriD just to be sure.

**Review by isapego — APPROVED** [review]
Overall looks good to me.

File: modules/platforms/dotnet/Apache.Ignite.Core/Ignition.cs (hunk 2 above)
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Ignition.cs:390]: Discussed in Slack, no objections from @slukyano and @DirectXceriD  .

**Pull request merged**